### PR TITLE
fix: add twisted edwards trait + proper constants

### PR DESCRIPTION
**Pull request by shuklaayush**

### CODE REVIEW

File: src/ed25519/curve.rs [hunk 1]
```diff
@@ -357,8 +345,13 @@ impl CurveExt for Ed25519 {
 
     const CURVE_ID: &'static str = "ed25519";
 
+    fn is_on_curve(&self) -> Choice {
+        let affine = Ed25519Affine::from(*self);
+        !self.z.is_zero() & affine.is_on_curve() & (affine.x * affine.y * self.z).ct_eq(&self.t)
```
Comment (jonathanpwang) [inline src/ed25519/curve.rs:354]: does this mean the infinity point will return false?

File: src/ed25519/fq.rs [hunk 2]
```diff
@@ -72,20 +72,34 @@ const TWO_INV: Fq = Fq::from_raw([
...
 ]);
 
-/// TODO
-const ZETA: Fq = Fq::zero();
-/// TODO
-const DELTA: Fq = Fq::zero();
-/// TODO
-const ROOT_OF_UNITY_INV: Fq = Fq::zero();
+// Element in small order subgroup (3-order)
+const ZETA: Fq = Fq::from_raw([
```
Comment (jonathanpwang) [inline src/ed25519/fq.rs:88]: could you add the sage commands to find these?

File: src/ed25519/fr.rs [hunk 3]
```diff
@@ -83,17 +87,34 @@ const TWO_INV: Fr = Fr::from_raw([
...
     0x0ef0565342ce83fe,
     0x7d3d6d60abc1c27a,
     0x094a7310e07981e7,
 ]);
 
-const ZETA: Fr = Fr::zero();
-const DELTA: Fr = Fr::zero();
-const ROOT_OF_UNITY_INV: Fr = Fr::zero();
+// Element in small order subgroup (3-order)
+const ZETA: Fr = Fr::from_raw([
```
Comment (jonathanpwang) [inline src/ed25519/fr.rs:103]: please provide sage commands if available

**Review by jonathanpwang — CHANGES_REQUESTED** [review]
A question about making sure to support is_on_curve for infinity point in CurveExt.

Also if you have the sage commands, it would be nice to include them for reproducibility.

File: src/ed25519/curve.rs (hunk 1 above)
Comment (shuklaayush) [reply src/ed25519/curve.rs:354]: No. There's no infinity point in TE curves. The identity element is `(x, y) = (0, 1)` which in [extended coordinates](https://en.wikipedia.org/wiki/Twisted_Edwards_curve#Extended_coordinates) corresponds to `(x, y, z, t) = (0, 1, 1, 0)`

File: src/ed25519/fq.rs (hunk 2 above)
Comment (shuklaayush) [reply src/ed25519/fq.rs:88]: Done

File: src/ed25519/fr.rs (hunk 3 above)
Comment (shuklaayush) [reply src/ed25519/fr.rs:103]: Done

File: src/ed25519/curve.rs (hunk 1 above)
Comment (jonathanpwang) [reply src/ed25519/curve.rs:354]: right... my bad

**Review by jonathanpwang — APPROVED** [review]
LGTM

**Comment by jonathanpwang** [comment]
@shuklaayush unfortunately there is some feature compile issue because of this `derive_serde` feature. Long term I feel like the `halo2-axiom` version can probably remove the feature and always have it enabled, but short term please do something like
```
#[cfg_attr(feature = "derive_serde", derive(Serialize, Deserialize))]
```

**Comment by shuklaayush** [comment]
@jonathanpwang Done, CI should pass now

**Pull request merged**